src/libDirectoryService/FinalBlockPreProcessing.cpp

```diff
@@ -184,8 +184,46 @@ bool DirectoryService::ComposeFinalBlock() {
       << m_mediator.m_txBlockChain.GetLastBlock().GetHeader().GetBlockNum() + 1
       << "][" << m_finalBlock->GetHeader().GetNumTxs() << "] FINAL");
 
-  LOG_EPOCH(INFO, m_mediator.m_currentEpochNum,
-            "Final block Composed: " << *m_finalBlock);
+  if (m_mediator.m_node->m_prePrepRunning) {
+    LOG_EPOCH(INFO, m_mediator.m_currentEpochNum,
+              "PrePrep Final block Composed: " << *m_finalBlock);
+  } else {
+    LOG_EPOCH(INFO, m_mediator.m_currentEpochNum,
+              "Final block Composed: " << *m_finalBlock);
+  }
+
+  m_completeFinalBlockReady = true;
+  m_cvCompleteFinalBlockReady.notify_all();
+
+  return true;
+}
+
+bool DirectoryService::WaitUntilCompleteFinalBlockIsReady() {
+  LOG_MARKER();
+  unique_lock<mutex> lock(m_mediator.m_node->m_mutexMicroBlock);
+  int timeout_time = std::max(
+      0, ((int)MICROBLOCK_TIMEOUT -
```

Comment:
This timeout definition seems a bit strange now because `MICROBLOCK_TIMEOUT` and `TX_DISTRIBUTE_TIME_IN_MS` have no more meaning for leader when it processes txns while the backups are confirming the pre-prepare txn list in the announcement message and then also processing the same txns on their end.

Reply:
My understanding is:
* `Node::NotifyTimeout` - actual timeout that cuts off txn processing in both leader and backup, in both DS- and shard-level consensus
* `Node::WaitUntilTxnProcessingDone` - timeout used by backup for MB readiness check, in both DS- and shard-level consensus
* `Node::WaitUntilCompleteMicroBlockIsReady` - timeout used by leader for MB readiness check, in shard-level consensus
* `DirectoryService::WaitUntilCompleteFinalBlockIsReady` - timeout used by leader for FB readiness check, in DS-level consensus

I wonder if we should change the timeout at shard- and DS-level consensus. Shard-level can probably use the current definition, but for DS-level maybe we have something that does not involve `MICROBLOCK_TIMEOUT` and `TX_DISTRIBUTE_TIME_IN_MS`?